File: src/clu/sources/no_salt.py
import logging

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_file

log = logging.getLogger(__name__)


class NoSalt(Source):
    def provides(self, provides: Provides) -> None:
        provides["salt.no_salt.exists"] = self
        provides["salt.no_salt.reason"] = self

    def requires(self, requires: Requires) -> None:
        requires.files.append("/no_salt")

    def parse(self, facts: Facts) -> None:
        data = text_file("/no_salt", optional=True)
        log.debug(f"{data=}")
        if data == "":
            facts["salt.no_salt.exists"] = "False"
            return
        else:
            log.debug(f"{data=}")
            facts["salt.no_salt.exists"] = "True"

        facts["salt.no_salt.reason"] = data.strip()

File: pkg/facts/sources/no_salt.go
package sources

// Detects presence of a sentinel file /no_salt and records existence and reason.

import (
	"strings"

	"github.com/huntermatthews/clu/pkg/facts/types"
	"github.com/huntermatthews/clu/pkg/input"
)

// NoSalt reports whether the /no_salt file exists and its content as a reason.
type NoSalt struct{}

var noSaltFacts = map[string]*types.Fact{
	"salt.no_salt.exists": {Name: "salt.no_salt.exists", Tier: types.TierOne},
	"salt.no_salt.reason": {Name: "salt.no_salt.reason", Tier: types.TierOne},
}

// Provides registers fact keys produced by this source.
func (n *NoSalt) Provides(p types.Provides) {
	for name := range noSaltFacts {
		p[name] = n
	}
}

// Requires declares file dependency.
func (n *NoSalt) Requires(r *types.Requires) {
	r.Files = append(r.Files, "/no_salt")
}

// Parse reads /no_salt optionally. If missing or empty sets exists False and returns.
// Otherwise sets exists True and reason to trimmed file content.
func (n *NoSalt) Parse(f *types.FactDB) {
	// Use DI FileReader (returns content,error). Treat any error or empty
	// trimmed content as missing (optional semantics).
	data, err := input.FileReader("/no_salt")
	if err != nil || strings.TrimSpace(data) == "" {
		noSaltFacts["salt.no_salt.exists"].Value = "False"
		f.AddFact(*noSaltFacts["salt.no_salt.exists"])
		return
	}
	noSaltFacts["salt.no_salt.exists"].Value = "True"
	noSaltFacts["salt.no_salt.reason"].Value = strings.TrimSpace(data)

	// Add all facts to the FactDB
	for _, fact := range noSaltFacts {
		f.AddFact(*fact)
	}
}
